borg/platform.py

```diff
@@ -14,3 +18,86 @@ def acl_get(path, item, st, numeric_owner=False):
 
     def acl_set(path, item, numeric_owner=False):
         pass
+
+# GenericDirEntry, scandir_generic (c) 2012 Ben Hoyt
+# from the python-scandir package (3-clause BSD license, just like us, so no troubles here)
+# note: simplified version
+
+
+class GenericDirEntry(object):
+    __slots__ = ('name', '_stat', '_lstat', '_scandir_path', '_path')
+
+    def __init__(self, scandir_path, name):
+        self._scandir_path = scandir_path
+        self.name = name
+        self._stat = None
+        self._lstat = None
+        self._path = None
+
+    @property
+    def path(self):
+        if self._path is None:
+            self._path = os.path.join(self._scandir_path, self.name)
+        return self._path
+
+    def stat(self, follow_symlinks=True):
+        assert not follow_symlinks
+        return os.lstat(self.path)
+
+    def is_dir(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink#
+        return stat.S_ISDIR(st.st_mode)
+
+    def is_file(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
+        return stat.S_ISREG(st.st_mode)
+
+    def is_symlink(self):
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
```

Comment:
Nope. this is lstat inside of self.stat(), so broken symlink is ok.
The only errors possible here are:
path does not exist because it was removed (ENOENT) OR
path does not eist because somebody is playing with us and replaced some intermediate component with a broken symlink or looping symlink (ENOENT or ELOOP) OR
somebody changed permissions along the pathname (I don't think you can change permissions on the symlink itself) (EACCESS) OR
somebody changed the path sucn that some intermediate component is no longer a directory (ENOTDIR) OR
something broke badly and you are getting EIO.

And then on top of all those, there are various resource constraints like ENOMEM